spec-tests/src/test/java/org/candlepin/spec/OwnerProductResourceSpecTest.java

```diff
@@ -653,82 +653,39 @@ public void shouldAllowDeletingAProductAssociatedWithAnActivationKey() {
         assertEquals(0, actual.getProducts().size());
     }
 
-    private static Stream<String> invalidStrings() {
-        return Stream.of("", "   ", null);
-    }
+    private static Stream<Arguments> criticalProductStringFieldsAndValues() {
+        Set<String> fields = Set.of("id", "name");
+        List<String> values = Arrays.asList("", "  ", null);
+        List<Arguments> matrix = new ArrayList<>();
 
-    @ParameterizedTest(name = "{displayName} {index}: {0}")
-    @MethodSource("invalidStrings")
-    public void shouldReturnBadRequestWhenInsertingProductWithInvalidName(String name) {
-        OwnerDTO owner = ownerApi.createOwner(Owners.random());
-        ProductDTO product = Products.random()
-            .name(name);
+        for (String field : fields) {
+            for (String value : values) {
+                matrix.add(Arguments.of(field, value));
+            }
+        }
 
-        assertBadRequest(() -> ownerProductApi.createProductByOwner(owner.getKey(), product));
+        return matrix.stream();
     }
 
-    @ParameterizedTest(name = "{displayName} {index}: {0}")
-    @MethodSource("invalidStrings")
-    public void shouldReturnBadRequestWhenInsertingProductWithInvalidId(String id) {
+    @ParameterizedTest(name = "{displayName} {index}: {0} {1}")
+    @MethodSource("criticalProductStringFieldsAndValues")
+    public void shouldRequireValidCriticalStringFieldsWhenInsertingProduct(String fieldName, String value)
+        throws Exception {
         OwnerDTO owner = ownerApi.createOwner(Owners.random());
-        ProductDTO product = Products.random()
-            .id(id);
-
-        assertBadRequest(() -> ownerProductApi.createProductByOwner(owner.getKey(), product));
-    }
-
-    @Nested
-    @TestInstance(Lifecycle.PER_CLASS)
-    public class JsonManipulationTests {
-        private ApiClient client;
-        private Gson unmodified;
-
-        @BeforeAll
-        public void beforeAll() {
-            // We need to enable serialization of nulls to craft the exact output we're expecting
-            // and ensuring we're actually testing fields that are explicitly set to null
-            this.client = ApiClients.admin();
-
-            this.unmodified = this.client.getApiClient().getJSON().getGson();
-            assertNotNull(this.unmodified);
-
-            Gson modified = this.unmodified.newBuilder()
-                .serializeNulls()
-                .create();
+        ObjectNode productNode = (ObjectNode) ApiClient.MAPPER.readTree(Contents.random().toJson());
```

Comment:
Couple things here:
1) this is building a content DTO from Contents.random(), rather than a product. Given the nature of this test, we should be manually building a fully-populated product so we can guarantee the target field is the one that's generating the error code we're expecting. Should also check the message includes the field in question.

2) The conversion to `ObjectNode` can be done such that you don't need a cast if you use `readValue` instead.

You can see examples of this in `OwnerContentSpecTest.shouldRequireCriticalStringFieldsAreValidWhenCreatingContent`

Reply:
Good catch!! 
- Updated the test to use a fully populated Product
- Updated to use `ApiClient.MAPPER.readValue` instead of `ApiClient.MAPPER.readTree`
- Added shouldRequirePopulatedCriticalStringFieldsWhenInsertingProduct to test that required fields are present